main.js

```diff
@@ -217,34 +216,108 @@ function updateDevices() {
         });
     });
 }
-// https://www.m-bus.de/vif.html
-const UNITS2ROLES = {
-    'W': 'value.power',
-    'kWh': 'value.power.consumption',
-    'kW': 'value.power',
-    '°C': 'value.temperature',
-    '°F': 'value.temperature',
-    '°K': 'value.temperature',
-    'm3/h': 'value.flow',
-    'm3/min': 'value.flow',
-    'm3/s': 'value.flow',
-    'm3': 'value.volume',
-    'sec': 'value.duration',
-    'hours': 'value.duration',
-    'minutes': 'value.duration'
-};
-
-const UNITS2UNITS = {
-    'deg C': '°C',
-    'deg K': '°K',
-    'deg F': '°F',
-    'seconds': 'sec',
-    'm m^3/h': 'm3/h',
-    'm m^3/min': 'm3/min',
-    'm m^3/s': 'm3/s',
-    'm^3': 'm3',
-    'm m^3': 'm3'
-};
+
+function adjustUnit(unit, type, forcekWh) {
+    let m;
+    // regex depending on type as to account for different units and keep it somewhat readable
+    switch (type) {
+        case "Energy": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(Wh|J)$/); break;
+        case "Mass": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(kg)$/); break;
+        case "Power": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(W|J\/h)$/); break;
+        case "Volume": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( m\^3)$/); break;
+        case "Volume flow": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( m\^3\/(h|min|s))$/); break;
+        case "Mass flow": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( kg\/h)$/); break;
+        case "Flow temperature":
+        case "Return temperature": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(deg (C|F))$/); break;
+        case "External temperature":
+        case "Temperature Difference": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( deg (C|F))$/); break;
+        case "Pressure":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( bar)$/); break;
+        case "Voltage":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( V)$/); break;
+        case "Current":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( A)$/); break;
+        case "Time Point": return {factor: undefined, unit: undefined};
+    }
+    //special case to adjust unit for durations
+    switch (unit) {
```

Comment:
why not move that up and do first? So the regex is not needed to be checked ...

Reply:
Hm. I think the way libmbus outputs data, there should never be a case where a case statement is true in both switch statements. So I guess it doesn't matter either way